src/bandersnatch_storage_plugins/filesystem.py

```diff
@@ -0,0 +1,248 @@
+        ) as f:
+            filepath_tmp = f.name
+            yield f
+
+        if not self.exists(filepath_tmp):
+            # Allow our clients to remove the file in case it doesn't want it to be
+            # put in place actually but also doesn't want to error out.
+            return
+        os.chmod(filepath_tmp, 0o100644)
+        self.copy_file(filepath_tmp, filepath)
+
+    @contextlib.contextmanager
+    def update_safe(self, filename: str, **kw: Any) -> Generator[IO, None, None]:
+        """Rewrite a file atomically.
+
+        Clients are allowed to delete the tmpfile to signal that they don't
+        want to have it updated.
+
+        """
+        with tempfile.NamedTemporaryFile(
+            dir=os.path.dirname(filename),
+            delete=False,
+            prefix=f"{os.path.basename(filename)}.",
+            **kw,
+        ) as tf:
+            if self.exists(filename):
+                os.chmod(tf.name, os.stat(filename).st_mode & 0o7777)
+            tf.has_changed = False  # type: ignore
+            yield tf
+            if not self.exists(tf.name):
+                return
+            filename_tmp = tf.name
+        if self.exists(filename) and self.compare_files(filename, filename_tmp):
+            os.unlink(filename_tmp)
+        else:
+            # TODO: account for other backends (add comparison functionality)
+            self.copy_file(filename_tmp, filename)
+            tf.has_changed = True  # type: ignore
+
+    def compare_files(self, file1: PATH_TYPES, file2: PATH_TYPES) -> bool:
+        """Compare two files, returning true if they are the same and False if not."""
+        return filecmp.cmp(str(file1), str(file2), shallow=False)
+
+    def copy_file(self, source: PATH_TYPES, dest: PATH_TYPES) -> None:
+        """Copy a file from **source** to **dest**"""
+        if not self.exists(source):
+            raise FileNotFoundError(source)
+        os.rename(source, dest)
+        return
+
+    def write_file(self, path: PATH_TYPES, contents: Union[str, bytes]) -> None:
+        """Write data to the provided path.  If **contents** is a string, the file will
+        be opened and written in "r" + "utf-8" mode, if bytes are supplied it will be
+        accessed using "rb" mode (i.e. binary write)."""
+        if not isinstance(path, pathlib.Path):
+            path = pathlib.Path(path)
+        if isinstance(contents, str):
+            path.write_text(contents)
+        else:
+            path.write_bytes(contents)
+
+    @overload
+    @contextlib.contextmanager
+    def open_file(self, path: PATH_TYPES, text: bool = True):
+        ...
+
+    @overload
+    @contextlib.contextmanager
+    def open_file(  # noqa
+        self, path: PATH_TYPES, text: bool = True, encoding: str = "utf-8"
+    ):
+        ...
+
+    @contextlib.contextmanager
+    def open_file(  # noqa
+        self, path: PATH_TYPES, text: bool = True, encoding: str = "utf-8"
```

Comment:
Why is this needed? Cause the args have defaults - Can't you just have the 1 definition? What's the mypy error?

Reply:
there is no error now apparently, i think i sorted it out :)